WIP: new prometheus middleware

aldas opened the pull request:
Proposal for new prometheus middleware (see https://github.com/labstack/echo-contrib/issues/80).  

This package 
* separates:
  * middleware (collects metrics), 
  * handler (serves collected metrics - pull) 
  * pushgateway (sends collected metrics - push) parts. 
*  Drops all `Metric` struct related features you are better off creating prometheus metrics (collectors) directly and not from indirections provided by old middleware.  This way we (maintainers) free ourselves of maintenance of these structs that build collectors. We hand full ownership of custom metrics to the user.
* New API does not start/run any goroutines - this is left to user. 
* Pushgateway related things are completely refactored. 

p.s. package name is `echoprometheus` because this way we get clean sheet and also it is quite annoying then you have conflicting package names (thing that we are using and middleware package). 

p.s.s. naming of things needs work.

Usage (shortest form):
```go
package main
import (
	"github.com/labstack/echo/v4"
	"github.com/labstack/echo-contrib/prometheus/echoprometheus"
)
	func main() {
	    e := echo.New()
	    // Enable metrics middleware
            e.Use(echoprometheus.NewMiddleware("myapp"))
            e.GET("/metrics", echoprometheus.NewHandler())
	    e.Logger.Fatal(e.Start(":1323"))
	}
```

Example for custom registry and custom metric usage:
```go
	customRegistry := prometheus.NewRegistry()
	customCounter := prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "custom_requests_total",
			Help: "How many HTTP requests processed, partitioned by status code and HTTP method.",
		},
	)
	if err := customRegistry.Register(customCounter); err != nil {
		t.Fatal(err)
	}

	e.Use(echoprometheus.NewMiddlewareWithConfig(echoprometheus.MiddlewareConfig{
		AfterNext: func(c echo.Context, err error) {
			customCounter.Inc() // use our custom metric in middleware
		},
		Registerer: customRegistry,
	}))
	e.GET("/metrics", echoprometheus.NewHandlerWithConfig(echoprometheus.HandlerConfig{Gatherer: customRegistry}))
```
that custom metric will appear in output as
```go
	assert.Contains(t, body, `custom_requests_total 1`)
```

Example for custom labels for middleware default metrics:
```go
	e.Use(echoprometheus.NewMiddlewareWithConfig(echoprometheus.MiddlewareConfig{
		LabelFuncs: map[string]echoprometheus.LabelValueFunc{
			"scheme": func(c echo.Context, err error) string { // additional custom label
				return c.Scheme()
			},
			"method": func(c echo.Context, err error) string { // overrides default 'method' label value
				return "overridden_" + c.Request().Method
			},
		},
	}))
	e.GET("/metrics", echoprometheus.NewHandler())
```

Example for changing middleware default metrics options before registering metrics:
```go
	e := echo.New()
	e.Use(echoprometheus.NewMiddlewareWithConfig(echoprometheus.MiddlewareConfig{
		HistogramOptsFunc: func(opts prometheus.HistogramOpts) prometheus.HistogramOpts {
			if opts.Name == "request_duration_seconds" {
				opts.ConstLabels = prometheus.Labels{"my_const": "123"}
				opts.Buckets = prometheus.DefBuckets
			}
			return opts
		},
	}))
```

Example running push gateway goroutine
```go
go func() {
	config := echoprometheus.PushGatewayConfig{
		PushGatewayURL: "https://host:9080",
		PushInterval:   10 * time.Millisecond,
	}
	if err := echoprometheus.RunPushGatewayGatherer(context.Background(), config); !errors.Is(err, context.Canceled) {
		log.Fatal(err)
	}
}()
```

aldas commented:
@lammel your thoughts?

yabberyabber commented:
The API you're putting forward looks really good to me. I especially like the way you expose `BeforeNext` and `AfterNext` which seems like an obvious hook with which to implement most types of data calculation.

I left a few nitty comments about implementation details although I have to say I'm very fond of the public API being exposed here. 

A few quick questions
 - Do you have a model in mind for whether/how the BeforeNext can create some context that gets referenced by the AfterNext? The example I have in mind is whether/how we could have implemented the `elapsed` time calculation outside of the API boundary. I see that there is already an elapsed time metric internally but I'm just curious to see whether something similar to this would be possible from outside the API boundary
 - Does the customMetrics API get a lot of usage out of customers right now? Are you aware of any specific customMetric implementations that need to be supported? Looking at the recent issues and PRs around it, I got the sense that it might not have been working until relatively recently. I'm wondering if maybe version 1 of this middleware aught to not support this functionality. Then the exact API can be developed later on once we have specific use-cases in mind.

lammel commented on prometheus/echoprometheus/prometheus.go:
```diff
@@ -0,0 +1,458 @@
+}
+
+// ToMiddleware converts configuration to middleware or returns an error.
+func (conf MiddlewareConfig) ToMiddleware() (echo.MiddlewareFunc, error) {
```
Does this have to be public? I guess most use cases are using NewMiddlewareWithConfig anyway

aldas replied:
This is to allowing to handle problems with middleware creation without panics.  Similar to https://github.com/labstack/echo-jwt/blob/95b0b607987a3bed484870b2052ef212763742a4/jwt.go#L173

lammel commented:
@aldas 

> About package name:
> 
> is good example why having middleware in `prometheus` package is annoying. When you need to access actual Prometheus stuff you have conflicting package names and need to alias which is inconvient and you still end up situation where you probably alias middleware package name to something other and leave Prometheus own package name as is.

Although aliasing for import works quite well I see your point. Conflicting package names can cause some headache.
But in that case I'd prefer to create a completely new echoprometheus middleware (not under `prometheus` directly in the echo-contrib root. That way it is quite clear, that no compatibility is expected, we can add short section "How to migrate from old prometheus middleware" to the README

> > Is the pushgateway functionality required (it is quite easy to setup a pushgateway which is normaly required for other metrics of the same node)
> 
> It is probably helpful.

Ok, keep it.

> 
> > which is normaly required for other metrics of the same node
> 
> I do not understand what you mean by that? In case you are creating custom collectors (metrics) for your background processes etc, you can register them with your `prometheus.Registerer` or `prometheus.DefaultRegistry` instance and pull/push them with your HTTP+GO etc related metrics.

I was referring to the deployment on servers, where a go daemon is usually just a small part of the puzzle. If push is required for metrics there is probably already a pushgateway running somewhere near to also push other node metrics (system metrics, other applications, ...). But as above, just keep it, as it's there already.

> 
> > Not sure if HistogramOptsFunc helper is required, 90% of the use case is probably to override bucketsize, but it's optional and very flexible
> 
> There are issues where people want to change buckets, namespace etc. It felt like most flexible way to allow users control over how default collectors/metrics are created. With that function you can change single one or all that pass through.

Understood, I agree this is the must flexible option we have.

> 
> > Should we switch to https://github.com/VictoriaMetrics/metrics if it's a new module anyway (this is somewhat offttopic, as we could simply add an echo-contrib/metrics package)
> 
> I think separate package would be better. Currently new API and old Prometheus MW API is atleast compatible when it comes to Prometheus own package structs/interfaces. Direct leap from Prometheus lib to VictoriaMetrics implementation is probably that takes time. Going from old API to new API is much easier when you do not do fancy things.

I completely agree.


So to **summarize**:
- All new `echoprometheus` middleware is the preferred way
- Just move from `prometheus/echoprometheus` to `echoprometheus`
- No compatibility to older prometheus middleware required or advertised
- No additional wrappers for other application metrics required (just use prometheus library)
- Short "How to migrate" section in a `echoprometheus/README.md` would be helpful

yabberyabber commented on prometheus/echoprometheus/prometheus.go:
```diff
@@ -0,0 +1,458 @@
+	if conf.Subsystem == "" {
+		conf.Subsystem = defaultSubsystem
+	}
+	if conf.Registerer == nil {
```
I see this pattern a lot and I'd like to propose an alternative - it involves adding this external library but the library has a pretty small dependency footprint https://github.com/imdario/mergo

Define a struct of defaults and merge it (the implementation uses reflect)
```
defaultConfig := MiddlewareConfig{
   Subsystem: defaultSubsystem,
   timeNow: time.Now,
   HistogramOptsFunc: func(blah blah blah),
}
err := mergo.Merge(&defaults, conf, mergo.WithOverride)
if err != nil { panic(err) }
conf = defaults
```

Things I find appealing about this are
 - it makes the code much briefer
 - you don't have to reason about what the zero-value is for anything because the type system (reflect package) handles that question
 - I find it easier to read


But I understand the resistance to adding dependencies as well as using reflect (which mergo uses heavily in its implementation)

aldas replied:
These public "defaults" objects are not that useful as looking middleware code separately it seems. 

I do not think I have ever felt the need to have these kind objects - most of the time you can get away just looking if field is nil or default value and replace it with something appropriate in middleware code. Only thing you would need is to fill field that interest you and need non-default values.  It looks eloquent but in real world it just adds complexity. Go is this stupidly simple language and things like that are somewhat against the Ethos of Go.

NB: not all code I produce follow the Ethos but I am simple man that often fails to see his own faults.

yabberyabber commented on prometheus/echoprometheus/prometheus.go:
```diff
@@ -0,0 +1,458 @@
+
+)
+
+	func main() {
```
The  indentation is inconsistent here

yabberyabber commented on prometheus/echoprometheus/prometheus.go:
```diff
@@ -0,0 +1,458 @@
+		return errors.New("push gateway URL is missing")
+	}
+	if config.PushInterval <= 0 {
+		config.PushInterval = 1 * time.Minute
```
Same comment here about mergo would apply if you choose to use mergo in this package

yabberyabber commented:
> I do not actually know what exactly are you meaning by `customMetrics API`. If you mean current `Metric` struct and functions related to that.
> 
> I do not think it should exists. Creation and usage of custom collector should be left to the developer. If we are talking about collectors/metrics related to HTTP requests then `BeforeNext` to `AfterNext` is the API which allows developer that freedom.

I guess what I'm trying to say is do we need to provide the `BeforeNext` and `AfterNext` hooks at all? Is there a specific behavior that we need to support that these enable?  My suggestion is maybe we omit them for now and wait to see who complains.

yabberyabber commented:
This API as written satisfies my usecase and I think is much easier to use than the previous API. Thanks for putting this together :+1:  Let me know if there's anything I can do on my end to help you get this in. I did a brief walk through and I only had a couple small comments.

yabberyabber commented:
Hey @aldas any update on this? I've got a project building off your fork and I would love to be able to point my go.mod replace rule at the upstream repo :)

Is there anything I can do to help this PR along?

aldas commented:
@yabberyabber If you are in a hurry and want to help - could you copy this middleware code at the moment in your application and start to use it - this would give us practical information how it works in real world.  If you are worried about licencing - this is MIT so just add the header and you are good.

I'll fix @lammel review comments and try to get this PR finished this week.

aldas commented:
@lammel please re-review. 

Mostly addresses these points from previous review
> * Just move from prometheus/echoprometheus to echoprometheus
> * Short "How to migrate" section in a echoprometheus/README.md would be helpful

lammel commented:
> @lammel please re-review.
> 
> Mostly addresses these points from previous review
> 
> > * Just move from prometheus/echoprometheus to echoprometheus
> > * Short "How to migrate" section in a echoprometheus/README.md would be helpful
> 
> and you know - I am no Shakespeare. so there is room for improvement in migration section.

Great! Just need some time to review. Give me a few days.

yabberyabber commented:
Thanks @aldas ! I just rebased my local projects to use the new middleware and it has been working great. 

I was just about to suggest you put the `Deprectaed` annotation on the old` prometheus` middleware but I see you've already done that :) That'll go a long way in helping current users to transition.

lammel commented on echoprometheus/README.md:
```diff
@@ -0,0 +1,154 @@
+`NewMetric` function allows you to create custom metrics to be registed with default Prometheus registry. Never middleware does
+not provide any abstractions for metrics and guides you to use native Prometheus metrics and register them by yourself.
+
+This can be done newer middleware as follows:
```
Suggestion: This can be done now as follows:

lammel commented on echoprometheus/README.md:
```diff
@@ -0,0 +1,154 @@
+
+## Replacement for `Metric.Buckets` and modifying default metrics
+
+Middleware register by default following metrics:
```
Suggestion:
The `echoprometheus` middleware registers the following metrics by default:

lammel commented on echoprometheus/README.md:
```diff
@@ -0,0 +1,154 @@
+`MetricsPath` was used to skip metrics own route from Prometheus metrics. Skipping is no longer done and requests to Prometheus
+route will be included in gathered metrics.
+
+This feature can be restored like that:
```
Suggestion:
To restore the old behaviour the `/metrics` path needs to be excluded from counting using the `Skipper` function:

lammel commented on echoprometheus/README.md:
```diff
@@ -0,0 +1,154 @@
+
+## Replacement for `Prometheus.RequestCounterURLLabelMappingFunc` and `Prometheus.RequestCounterHostLabelMappingFunc`
+
+These function fields were used to define how "URL" or "Host" attribute in Prometheus metric lines were created.
```
Typo: ... lines **are** created.

These can **now** be substituted ...

lammel reviewed (changes requested):
Looks good! 
Only some minor textual suggestions.

aldas commented:
all right, textual changes done. @lammel, your approvement please and I'll merge and create next release.

lammel reviewed (approved):
LGTM!

aldas commented:
Thanks  @yabberyabber and @lammel. I'll update Echo website soon
